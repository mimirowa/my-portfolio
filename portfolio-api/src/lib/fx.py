from datetime import date as date_cls
from typing import Union
import os

import requests
from flask import current_app

from src.models.portfolio import FxRate
from src.models.user import db


class UnsupportedCurrency(Exception):
    pass


class FxDownloadError(Exception):
    pass


def _norm(code: str) -> str:
    code = code.strip().upper()
    if len(code) != 3 or not code.isascii() or not code.isalpha():
        raise UnsupportedCurrency(code)
    return code


validate_currency_code = _norm


def get_fx_rate(from_ccy: str, to_ccy: str, dt: Union[str, date_cls]) -> float:
    """Return FX rate from ``from_ccy`` to ``to_ccy`` for ``dt``.

    Checks local cache first and falls back to AlphaVantage FX_DAILY.
    """
    if isinstance(dt, str):
        dt = date_cls.fromisoformat(dt)

    from_ccy = _norm(from_ccy)
    to_ccy = _norm(to_ccy)

    if from_ccy == to_ccy:
        return 1.0

    rate = FxRate.query.filter_by(base=from_ccy, target=to_ccy, date=dt).first()
    if rate:
        return rate.rate

    api_key = os.environ.get("ALPHAVANTAGE_API_KEY")
    if not api_key:
        raise FxDownloadError("missing API key")

    params = {
        "function": "FX_DAILY",
        "from_symbol": from_ccy,
        "to_symbol": to_ccy,
        "outputsize": "compact",
        "apikey": api_key,
    }
    try:
        resp = requests.get("https://www.alphavantage.co/query", params=params, timeout=10)
        data = resp.json()
        if "Note" in data:
            raise FxDownloadError("quota exceeded")
        if "Error Message" in data:
            raise FxDownloadError("invalid pair")
        series = data["Time Series FX (Daily)"]
        fx = float(series[dt.isoformat()]["4. close"])
    except FxDownloadError:
        raise
    except Exception as exc:
        current_app.logger.exception("FX fetch failed: %s", exc)
        raise FxDownloadError("unreachable") from exc

    db.session.add(FxRate(base=from_ccy, target=to_ccy, date=dt, rate=fx))
    db.session.commit()

    return fx
