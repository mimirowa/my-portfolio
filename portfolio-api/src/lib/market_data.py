import os
import requests


_API_URL = "https://www.alphavantage.co/query"


def _fetch_av_overview(symbol: str, api_key: str) -> str | None:
    """Return company name using Alpha Vantage OVERVIEW."""
    try:
        resp = requests.get(
            _API_URL,
            params={"function": "OVERVIEW", "symbol": symbol, "apikey": api_key},
            timeout=10,
        )
        resp.raise_for_status()
        data = resp.json()
        return data.get("Name") or None
    except Exception:
        return None


def _fetch_yahoo_company(symbol: str) -> str | None:
    """Return company name using Yahoo Finance quoteSummary."""
    url = f"https://query1.finance.yahoo.com/v10/finance/quoteSummary/{symbol}"
    try:
        resp = requests.get(url, params={"modules": "price"}, timeout=10)
        resp.raise_for_status()
        data = resp.json()
        result = data.get("quoteSummary", {}).get("result")
        if result:
            price_info = result[0].get("price", {})
            return price_info.get("longName") or price_info.get("shortName")
    except Exception:
        return None
    return None


def get_company_name(ticker: str) -> str | None:
    """Return company name for ticker symbol using Alpha Vantage or Yahoo."""
    symbol = ticker.upper()
    api_key = os.environ.get("ALPHAVANTAGE_API_KEY")

    if api_key:
        name = _fetch_av_overview(symbol, api_key)
        if name:
            return name

    return _fetch_yahoo_company(symbol)
