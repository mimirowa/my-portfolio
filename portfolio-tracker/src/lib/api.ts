const BASE = (import.meta?.env?.VITE_API_URL || '').replace(/\/portfolio$/, '')

export async function parseGoogleFinance(raw: string) {
  const resp = await fetch(`${BASE}/import/google-finance/preview`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ raw })
  })
  if (!resp.ok) throw new Error('Failed to parse')
  return resp.json()
}

export async function importGoogleFinance(raw: string) {
  const resp = await fetch(`${BASE}/import/google-finance`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ raw })
  })
  if (!resp.ok) throw new Error('Failed to import')
  return resp.json()
}
