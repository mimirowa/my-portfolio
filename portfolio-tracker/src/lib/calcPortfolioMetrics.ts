export interface Holding {
  symbol: string
  company_name?: string
  quantity: number
  avg_cost_basis: number
  current_price?: number | null
  fees_paid?: number
}

export interface HoldingMetrics extends Holding {
  current_value: number
  cost_basis: number
  total_gain: number
  total_gain_percent: number
}

export interface PortfolioMetrics {
  total_value: number
  total_cost_basis: number
  total_gain: number
  total_gain_percent: number
  total_fees_paid: number
  net_gain_after_fees: number
  stocks_count: number
  holdings: HoldingMetrics[]
}

export function calcPortfolioMetrics(holdings: Holding[]): PortfolioMetrics {
  let totalValue = 0
  let totalCostBasis = 0
  let totalFees = 0
  const metrics: HoldingMetrics[] = holdings.map((h) => {
    const price = h.current_price ?? 0
    const currentValue = h.quantity * price
    const costBasis = h.quantity * h.avg_cost_basis
    const totalGain = currentValue - costBasis
    const totalGainPercent = costBasis > 0 ? (totalGain / costBasis) * 100 : 0
    totalValue += currentValue
    totalCostBasis += costBasis
    totalFees += h.fees_paid ?? 0
    return {
      ...h,
      current_value: currentValue,
      cost_basis: costBasis,
      total_gain: totalGain,
      total_gain_percent: totalGainPercent,
    }
  })
  const totalGain = totalValue - totalCostBasis
  const totalGainPercent =
    totalCostBasis > 0 ? (totalGain / totalCostBasis) * 100 : 0
  return {
    total_value: totalValue,
    total_cost_basis: totalCostBasis,
    total_gain: totalGain,
    total_gain_percent: totalGainPercent,
    total_fees_paid: totalFees,
    net_gain_after_fees: totalGain - totalFees,
    stocks_count: metrics.length,
    holdings: metrics,
  }
}
