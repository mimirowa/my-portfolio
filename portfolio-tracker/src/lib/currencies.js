export const SUPPORTED_CCY = ['USD','EUR','GBP','SEK','PLN','JPY']

