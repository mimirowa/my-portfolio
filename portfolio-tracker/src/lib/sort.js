export function sortPerformanceData(data, metric = 'current_value') {
  const key = metric === 'total_gain' ? 'gain' : 'currentValue'
  return [...data].sort((a, b) => b[key] - a[key])
}
