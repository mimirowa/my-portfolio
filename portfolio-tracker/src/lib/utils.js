import { clsx } from "clsx";
import { twMerge } from "tailwind-merge"

export function cn(...inputs) {
  return twMerge(clsx(inputs));
}

export function getCurrencySymbol(code) {
  const symbols = {
    USD: '$',
    EUR: '€',
    SEK: 'kr',
    GBP: '£',
    JPY: '¥'
  }
  return symbols[code] || code
}
