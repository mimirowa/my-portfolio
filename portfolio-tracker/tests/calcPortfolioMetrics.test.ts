import { calcPortfolioMetrics } from '../src/lib/calcPortfolioMetrics'

test('calcPortfolioMetrics computes totals', () => {
  const holdings = [
    { symbol: 'AAPL', quantity: 10, avg_cost_basis: 100, current_price: 150 },
    { symbol: 'MSFT', quantity: 5, avg_cost_basis: 200, current_price: 180 },
  ]
  const result = calcPortfolioMetrics(holdings)
  expect(result).toMatchSnapshot()
})
