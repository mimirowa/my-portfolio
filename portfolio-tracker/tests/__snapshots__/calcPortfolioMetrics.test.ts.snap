// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`calcPortfolioMetrics computes totals 1`] = `
{
  "holdings": [
    {
      "avg_cost_basis": 100,
      "cost_basis": 1000,
      "current_price": 150,
      "current_value": 1500,
      "quantity": 10,
      "symbol": "AAPL",
      "total_gain": 500,
      "total_gain_percent": 50,
    },
    {
      "avg_cost_basis": 200,
      "cost_basis": 1000,
      "current_price": 180,
      "current_value": 900,
      "quantity": 5,
      "symbol": "MSFT",
      "total_gain": -100,
      "total_gain_percent": -10,
    },
  ],
  "stocks_count": 2,
  "total_cost_basis": 2000,
  "total_gain": 400,
  "total_gain_percent": 20,
  "total_value": 2400,
}
`;
